### PR TITLE
test: setup out-of-cluster test execution in pipeline

### PR DESCRIPTION
**Pull request by yangchiu**
test: setup out-of-cluster test execution in pipeline

For https://github.com/longhorn/longhorn/issues/6667

Signed-off-by: Yang Chiu <yang.chiu@suse.com>

### CODE REVIEW

**Comment by yangchiu** [comment]
> What’s the cost if running a week?

For t2.xlarge instances, it would be: 0.1856×4×24×7=124.7232

**Comment by innobead** [comment]
Thanks. For the next, we should consider running this in the local lab in the near future. Please discuss with Khushboo later and plan the tasks accordingly.

File: e2e/libs/node/node.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ def reboot_node(self, running_on_node_name, reboot_node_name, shut_down_time_in_
             mapping = yaml.safe_load(f)
             reboot_node_id = mapping[reboot_node_name]
 
-        filepath = './litmus/reboot-node.yaml'
+        filepath = './templates/litmus/reboot-node.yaml'
```
Comment (innobead) [inline e2e/libs/node/node.py:41]: Suggest having:
- a template util function to get the file
- use const variable for any constant file path like `tmp/instance_mapping`

You can do it later over time.

File: pipelines/e2e/Dockerfile.setup [hunk 2]
```diff
@@ -25,7 +25,7 @@ RUN wget -q https://storage.googleapis.com/kubernetes-release/release/$KUBECTL_V
     wget -q "https://github.com/mikefarah/yq/releases/download/${YQ_VERSION}/yq_linux_amd64" && \
     mv yq_linux_amd64 /usr/local/bin/yq && \
     chmod +x /usr/local/bin/yq && \
-    apk add openssl openssh-client ca-certificates git rsync bash curl jq python3 py3-pip && \
+    apk add openssl openssh-client ca-certificates git rsync bash curl jq python3 py3-pip gcc python3-dev libc-dev && \
```
Comment (innobead) [inline pipelines/e2e/Dockerfile.setup:28]: What's purpose to have gcc and libc here?

File: pipelines/utilities/run_longhorn_e2e_test.sh [hunk 3]
```diff
@@ -63,3 +63,16 @@ run_longhorn_e2e_test(){
   kubectl cp ${LONGHORN_TEST_POD_NAME}:/tmp/test-report/output.xml "output.xml" -c longhorn-test-report
   kubectl cp ${LONGHORN_TEST_POD_NAME}:/tmp/test-report/report.html "report.html" -c longhorn-test-report
 }
+
+run_longhorn_e2e_test_out_of_cluster(){
+  cd e2e
+  pip install -r requirements.txt
```
Comment (innobead) [inline pipelines/utilities/run_longhorn_e2e_test.sh:69]: If running test cases, out of cluster, we should still use container to run it independent of the host. You can assume docker here or any container manager CLI (nerdctl) you prefer.

File: pipelines/utilities/terraform_setup.sh [hunk 4]
```diff
@@ -24,4 +24,6 @@ if [[ "${TF_VAR_k8s_distro_name}" == "k3s" ]]; then
   terraform -chdir=test_framework/terraform/${LONGHORN_TEST_CLOUDPROVIDER}/${DISTRO} output -raw instance_mapping | jq 'map({(.name | split(".")[0]): .id}) | add' | jq -s add > /tmp/instance_mapping
 fi
 
+terraform -chdir=test_framework/terraform/${LONGHORN_TEST_CLOUDPROVIDER}/${DISTRO} output -raw controlplane_public_ip > /tmp/controlplane_public_ip
```
Comment (innobead) [inline pipelines/utilities/terraform_setup.sh:27]: Where `/tmp/controlplane_public_ip` will be used?

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Some comments need to resolve first.

File: pipelines/e2e/Dockerfile.setup (hunk 2 above)
Comment (yangchiu) [reply pipelines/e2e/Dockerfile.setup:28]: To satisfy package installation in `requirements.txt`.

File: pipelines/utilities/run_longhorn_e2e_test.sh (hunk 3 above)
Comment (yangchiu) [reply pipelines/utilities/run_longhorn_e2e_test.sh:69]: It already runs in a container. Reference https://github.com/longhorn/longhorn-tests/blob/master/pipelines/e2e/Jenkinsfile#L75 for more details.

File: pipelines/utilities/terraform_setup.sh (hunk 4 above)
Comment (yangchiu) [reply pipelines/utilities/terraform_setup.sh:27]: For longhorn api client: https://github.com/longhorn/longhorn-tests/pull/1530/files#diff-c3e9bb86df59b862e75a6e5779a1f6631e809616c9de781e9a58d2b249d4dc06

File: pipelines/e2e/Dockerfile.setup (hunk 2 above)
Comment (innobead) [reply pipelines/e2e/Dockerfile.setup:28]: Do we know which package requires those?

File: e2e/libs/node/node.py (hunk 1 above)
Comment (yangchiu) [reply e2e/libs/node/node.py:41]: Created https://github.com/longhorn/longhorn-tests/pull/1530#discussion_r1328727134 for this.

File: pipelines/e2e/Dockerfile.setup (hunk 2 above)
Comment (yangchiu) [reply pipelines/e2e/Dockerfile.setup:28]:
```
# pip install directio==1.2
Collecting directio==1.2
  Using cached directio-1.2.tar.gz (3.1 kB)
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: directio
  Building wheel for directio (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [9 lines of output]
      Warning: 'classifiers' should be a list, got type 'filter'
      running bdist_wheel
      running build
      running build_ext
      building 'directio' extension
      creating build
      creating build/temp.linux-x86_64-3.10
      gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fomit-frame-pointer -g -fomit-frame-pointer -g -fomit-frame-pointer -g -DTHREAD_STACK_SIZE=0x100000 -fPIC -I/usr/include/python3.10 -c directiomodule.c -o build/temp.linux-x86_64-3.10/directiomodule.o
      error: command 'gcc' failed: No such file or directory
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for directio
  Running setup.py clean for directio
Failed to build directio
Installing collected packages: directio
  Running setup.py install for directio ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for directio did not run successfully.
  │ exit code: 1
  ╰─> [11 lines of output]
      Warning: 'classifiers' should be a list, got type 'filter'
      running install
      /usr/lib/python3.10/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_ext
      building 'directio' extension
      creating build
      creating build/temp.linux-x86_64-3.10
      gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fomit-frame-pointer -g -fomit-frame-pointer -g -fomit-frame-pointer -g -DTHREAD_STACK_SIZE=0x100000 -fPIC -I/usr/include/python3.10 -c directiomodule.c -o build/temp.linux-x86_64-3.10/directiomodule.o
      error: command 'gcc' failed: No such file or directory
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> directio

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
```

File: e2e/libs/node/node.py (hunk 1 above)
Comment (innobead) [reply e2e/libs/node/node.py:41]: @yangchiu which issue you have been created?

Comment (yangchiu) [reply e2e/libs/node/node.py:41]: Sorry for the wrong link.. Updated!

**Comment by yangchiu** [comment]
> Thanks. For the next, we should consider running this in the local lab in the near future. Please discuss with Khushboo later and plan the tasks accordingly.

These test cases are not applicable in local lab. FYI

Since currently we use `aws client` to shutdown/power off an instance.

Implementation from scratch is required to run them in local lab.

For `reboot` test cases, `ssh` approach should be implemented to execute `shutdown -r now` on nodes.

For `power off X minutes` test cases, even `ssh` can't achieve this operation, since we are unable to `power on` the node through `ssh`. How to `power on` a node or VM in this local lab is unknown.

**Pull request merged**